docs(website): add docs for code-editor(-library) and vs-code-plugin (#3332)
twilio-labs/paste: packages/paste-website/src/pages/components/editable-code-block/index.mdx

  data={props.data}
/>

---

<contentwrapper>

<PageAside data={mdxHeadings} />

<content>

<StoryPreview
  title="Default Editable Code Block"
  storyID="components-editable-code-block--default"
  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>Typescript Example</EditableCodeBlockHeader>
  <EditableCodeBlock
    height="45vh"
    defaultLanguage="typescript"
    defaultValue={TypeScriptExample}
  />
</EditableCodeBlockWrapper>`}
/>

## Guidelines

### About Editable Code Block

Editable Code Block allows basic text input features for code-editing experiences on the
web. To help users differentiate this from a static [Code Block](/components/code-block),
Editable Code Block enables line numbers, code folding, and indentation guides by default.

This component is built on top of the Paste [Code Editor library](/core/libraries/code-editor),
which wraps the [Monaco Editor](https://microsoft.github.io/monaco-editor/) used by Visual
Studio Code. If you’re looking for more functionality than what’s provided through Editable
Code Block, use the Code Editor library instead.

> This component uses the [Night Owl theme](https://marketplace.visualstudio.com/items?itemName=sdras.night-owl),
> an accessible theme by Sarah Drasner for people with colorblindness and low vision situations.
> If you’re looking for more functionality than what’s provided through Editable Code Block,
> use the [Code Editor library](/core/libraries/code-editor).

Reviewer: ```suggestion
This component uses the [Night Owl theme](https://marketplace.visualstudio.com/items?itemName=sdras.night-owl),
an accessible theme by Sarah Drasner for people with colorblindness and low vision situations.
```

    inlineErrorHoverMessage={{value: '"id" can only be a string type.'}}
    defaultLanguage="typescript"
    defaultValue={TypeScriptExample}
  />
</EditableCodeBlockWrapper>`}
/>

### Generic errors

If there’s an error that can’t be traced to specific line numbers, use a [Callout](/components/callout).

For additional guidance on how to compose error messages, refer to the [error state pattern](/patterns/error-state).

<StoryPreview
  title="Generic errors in Editable Code Block"
  storyID="components-editable-code-block--generic-error"
  code={`<Stack orientation="vertical" spacing="space40">
  <Callout variant="error">
    <CalloutHeading as="h3">401 Unauthorized Request</CalloutHeading>
    <CalloutText>Please provide a valid API key.</CalloutText>
  </Callout>
  <EditableCodeBlockWrapper>
    <EditableCodeBlockHeader>Generic error example</EditableCodeBlockHeader>
    <EditableCodeBlock height="45vh" defaultLanguage="typescript" defaultValue="const API_KEY = null;" />
  </EditableCodeBlockWrapper>
</Stack>`}
/>

### Read only

Only if the user doesn't have permission to edit the code, use the `readOnly` prop to
make the Editable Code Block read-only. In most cases, use the

> [Code Block](/components/code-block] component instead to show static code.

Reviewer: ```suggestion
[Code Block](/components/code-block) component instead to show static code.
```

sorry, typo'd in my original suggestion here.

  slug: '/components/editable-code-block/',
};

import Changelog from '@twilio-paste/editable-code-block/CHANGELOG.md';
import packageJson from '@twilio-paste/editable-code-block/package.json';

import {DoDont, Do, Dont} from '../../../components/DoDont';
import {Blockquote} from '../../../components/Blockquote';
import {Codeblock} from '../../../components/codeblock';
import {SidebarCategoryRoutes} from '../../../constants';
import DefaultLayout from '../../../layouts/DefaultLayout';
import {getFeature, getNavigationData} from '../../../utils/api';

export default DefaultLayout;

export const getStaticProps = async () => {
  const navigationData = await getNavigationData();
  const feature = await getFeature('Editable Code Block');
  return {
    props: {
      data: {
        ...packageJson,
        ...feature,
      },
      navigationData,
    },
  };
};

<NormalizedComponentHeader
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  storybookUrl="/?path=/story/components-editable-code-block--"
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/editable-code-block"
  data={props.data}

> />

Author: Fixed

---

<contentwrapper>

<PageAside data={mdxHeadings} />

<content>

> <StoryPreview

Reviewer 2: non-blocking: there's a ton of whitespace around the codeblock in these previews, is there anything we can do about that?

Author: Can't figure it out easily. Removing the padding only covers top left and right, but the border edges still peek through and there main issue is bottom whitespace due to not being able to determine height correctly
<img width="776" alt="image" src="https://github.com/twilio-labs/paste/assets/1592327/bb5fc0d1-05f2-4ca2-a66c-1476c031cf15">
Just going to leave it as-is for now

  title="Default Editable Code Block"
  storyID="components-editable-code-block--default"
  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>Typescript Example</EditableCodeBlockHeader>
  <EditableCodeBlock
    height="45vh"
    defaultLanguage="typescript"
    defaultValue={TypeScriptExample}
  />
</EditableCodeBlockWrapper>`}
/>

## Guidelines

### About Editable Code Block

> Editable Code Block allows basic text input features for code-editing experiences on web,
> including syntax highlighting and displaying line numbers.

Reviewer: ```suggestion
Editable Code Block has basic text input features for code editing experiences on the web. To help users differentiate this from a static [[Code Block](https://paste.twilio.design/components/code-block)](/components/code-block), Editable Code Block enables line numbers, code folding, and indentation guides by default.
```

This component is built on top of the [Code Editor library](/core/libraries/code-editor),

> which wraps the [Monaco Editor](https://microsoft.github.io/monaco-editor/) used by Visual Studio Code.
> 
> This component uses the [Night Owl theme](https://marketplace.visualstudio.com/items?itemName=sdras.night-owl).
> If you’re looking for more functionality than what’s provided through Editable Code Block, use the [Code Editor library](/core/libraries/code-editor).

Reviewer: ```suggestion
which wraps the [Monaco Editor](https://microsoft.github.io/monaco-editor/) used by Visual Studio Code. If you’re looking for more functionality than what’s provided through Editable Code Block, use the Code Editor library instead.

This component uses the [Night Owl theme](https://marketplace.visualstudio.com/items?itemName=sdras.night-owl), an accessible theme by Sarah Drasner for people with colorblindness and low vision situations.
```

### Accessibility

Because Editable Code Block takes text input, it follows similar labeling accessibility
guidelines as other form elements:

- Include a label on _all_ Editable Code Blocks. Use one of these ways to label an Editable Code Block:

>   - Editable Code Block Header (preferred). Use the correct heading level in Editable Code Block Header.

Reviewer: ```suggestion
  - [Editable Code Block Header](#default) (preferred). Use the correct heading level in Editable Code Block Header.
```

  - Visible label (for example, a page heading) that's associated to the input with `aria-labelledby`
  - Visible label with [Label](/components/label)
  - Label directly using `aria-label`
- Each label must use the `htmlFor` prop that equals the `id` of the associated Input.

Since Editable Code Block takes up more vertical space than typical form elements,
any help or error content should appear before the component by using an

> introductory [Paragraph](/components/paragraph) or [error Callout](#Generic-errors).

Reviewer: URL should also be `#generic-errors`

#### Keyboard interaction

> Excerpts from [Monaco accessibility guide](https://github.com/Microsoft/monaco-editor/wiki/Monaco-Editor-Accessibility-Guide):

Reviewer: ```suggestion
Excerpts from the [Monaco accessibility guide](https://github.com/Microsoft/monaco-editor/wiki/Monaco-Editor-Accessibility-Guide):
```

| Toggle tab trapping             | Ctrl+Shift+M or Ctrl+M        | “Subsequent Tab keys will move focus out of the editor.”                                                                                                                                                      |
| Go to Next Error or Warning     | F8                            |                                                                                                                                                                                                               |
| Go to Previous Error or Warning | Shift+F8                      |                                                                                                                                                                                                               |
| Show accessibility help         | Option+F1, Alt+F1, or Ctrl+F1 | Shows a “dialog…to find out the current position in the editor and to check the state of various accessibility options. The editor can be dynamically optimized for screen reader software from this dialog.” |

## Examples

### Default

> Use an Editable Code Block to allow users to enter code on multiple lines.
> To help users differentiate this from a static [Code Block](/components/code-block),
> Editable Code Block enables line numbers, code folding, and indentation guides by default.
> 
> Put a logical label in an _Editable Code Block Header_ to describe what users need
> to put in the code block and/or the language used, similar to a form label.
> 
> _Line numbers_ help users identify errors and cross-reference with existing code.
> Use line numbers when the expected input is longer than 5 lines.
> 
> _Indentation guides and code folding_ help users read long blocks of
> code and troubleshoot errors.

Reviewer: ```suggestion
The default Editable Code Block includes:

- An Editable Code Block Header: Contains a logical label at the correct heading level that describes what users need to put in the code block and/or the language used.
- Line numbers: Helps users identify errors and cross-reference with existing code. Use line numbers when the expected input is longer than 5 lines.
- Indentation guides and code folding: Helps users read long blocks of code and troubleshoot errors.
```

<StoryPreview
  title="Default Editable Code Block"
  storyID="components-editable-code-block--default"
  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>Typescript Example</EditableCodeBlockHeader>
  <EditableCodeBlock
    height="45vh"
    defaultLanguage="typescript"
    defaultValue={TypeScriptExample}
  />
</EditableCodeBlockWrapper>`}
/>

### Simple Editable Code Block

Remove line numbers, indentation guides, and code folding from Editable Code Block only when:

- Your page layout makes it visually clear to the user that the code block is editable, _and_
- The expected code input is simple, like a short snippet of JSON.

<StoryPreview
  title="Simple Editable Code Block"

>   storyID="components-editable-code-block--no-special-features"

Reviewer 2: issue: this preview and some other aren't working
![image](https://github.com/twilio-labs/paste/assets/75342690/4acf4cc6-67af-49ee-aefe-cddc5d8e4c69)

Author: This new storybook preview component only works after merge since it pulls the stories from the main branch storybook deploy

  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>Simple Editable Code Block</EditableCodeBlockHeader>
  <EditableCodeBlock
    height="45vh"
    defaultLanguage="typescript"
    lineNumbers="off"
    folding={false}
    indentationGuide={false}
    defaultValue={TypeScriptExample}
  />
</EditableCodeBlockWrapper>`}
/>

### Editable Code Block with minimap

For large files, use a minimap to help users navigate the code block.

<StoryPreview
  title="Editable Code Block with minimap"
  storyID="components-editable-code-block--minimap"
  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>Minimap example</EditableCodeBlockHeader>
  <EditableCodeBlock height="45vh" defaultLanguage="typescript" showMinimap defaultValue={TypeScriptExample} />
</EditableCodeBlockWrapper>`}
/>

### Multi-file Editable Code Block

> You can alternate the content in a single instance of the Editable Code Block.

Reviewer 2: ```suggestion
You can alternate the content in a single instance of the Editable Code Block using [Radio Button Group](/components/radio-button-group).
```

This is useful for editing multiple files in a single editor.

<StoryPreview
  title="Multi-file Editable Code Block"
  storyID="components-editable-code-block--multi-file-editor"
  code={`<Stack orientation="vertical" spacing="space40">
  <RadioButtonGroup
    attached
    name="file-selector"
    legend="Select file"
    helpText="Show code for the selected file"
    onChange={(_fileKey) => setFileKey(_fileKey)}
  >
    {Object.keys(Files).map((key) => (
      <RadioButton key={key} value={key} defaultChecked={fileKey === key}>
        {Files[key].name}
      </RadioButton>
    ))}
  </RadioButtonGroup>
  <EditableCodeBlockWrapper>
    <EditableCodeBlockHeader>{Files[fileKey].name}</EditableCodeBlockHeader>
    <EditableCodeBlock
      height="400px"
      path={Files[fileKey].name}
      defaultLanguage={Files[fileKey].language}
      defaultValue={Files[fileKey].value}
    />
  </EditableCodeBlockWrapper>
</Stack>`}

> />

Reviewer: suggestion: Remove this example. We didn't design for this case, and it'll be better if people tell us they need it, so we can add tabs appropriately to the component, instead of them trying to use this as is.

## States

### Read only

> If the user doesn't have permission to edit the code, use the `readOnly` prop to make the Editable Code Block read-only.

Reviewer: ```suggestion
Only if the user doesn't have permission to edit the code, use the `readOnly` prop to make the Editable Code Block read-only. In most cases, use the [Code Block](/components/code-block] component instead to show static code. 
```

<StoryPreview
  title="Read-Only Editable Code Block"
  storyID="components-editable-code-block--read-only"
  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>ReadOnly Example</EditableCodeBlockHeader>
  <EditableCodeBlock height="45vh" readOnly defaultLanguage="typescript" defaultValue={TypeScriptExample} />
</EditableCodeBlockWrapper>`}

> />

Reviewer: suggestion: Move this underneath the error examples since it's a less-common use case

### Syntax errors

Monaco Editor will detect and show syntax errors out of the box.

<StoryPreview
  title="Syntax error example in Editable Code Block"
  storyID="components-editable-code-block--syntax-error"
  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>Syntax error example</EditableCodeBlockHeader>
  <EditableCodeBlock
    height="45vh"
    defaultLanguage="typescript"
    defaultValue={CodeStringWithSyntaxError}
  />
</EditableCodeBlockWrapper>`}
/>

> ### Line-specific errors

Reviewer: ```suggestion
### Custom line-specific errors
```

> Show line-specific errors to help users pinpoint what line of code they need to fix.

Reviewer: ```suggestion
Show line-specific errors to help users pinpoint what line of code they need to fix.

For additional guidance on how to compose error messages, refer to the [error state pattern](/patterns/error-state).
```

<StoryPreview
  title="Custom line-specific errors in Editable Code Block"
  storyID="components-editable-code-block--custom-inline-error"
  code={`<EditableCodeBlockWrapper>
  <EditableCodeBlockHeader>Custom inline error example (try hovering it)</EditableCodeBlockHeader>
  <EditableCodeBlock
    height="45vh"
    onChange={(value) => console.log(value)}
    inlineErrorRange={{
      startLineNumber: 3,
      endLineNumber: 3,
      startColumn: 7,
      endColumn: 13,
    }}
    inlineErrorIsWholeLine={false}
    inlineErrorHoverMessage={{value: '"id" can only be a string type.'}}
    defaultLanguage="typescript"
    defaultValue={TypeScriptExample}
  />
</EditableCodeBlockWrapper>`}
/>

### Generic errors

If there’s an error that can’t be traced to specific line numbers, use a [Callout](/components/callout).

> For additional guidance on how to compose error messages, refer to the [error state pattern](/patterns/error-state).

Reviewer 2: ```suggestion
For additional guidance on how to compose error messages, refer to the [Error State Pattern](/patterns/error-state).
```

Reviewer: We keep patterns lowercase, generally, to help differentiate them from components which are like built Things

Reviewer 2: news to me lol

<StoryPreview
  title="Generic errors in Editable Code Block"
  storyID="components-editable-code-block--generic-error"
  code={`<Stack orientation="vertical" spacing="space40">
  <Callout variant="error">
    <CalloutHeading as="h3">401 Unauthorized Request</CalloutHeading>
    <CalloutText>Please provide a valid API key.</CalloutText>
  </Callout>
  <EditableCodeBlockWrapper>
    <EditableCodeBlockHeader>Generic error example</EditableCodeBlockHeader>
    <EditableCodeBlock height="45vh" defaultLanguage="typescript" defaultValue="const API_KEY = null;" />
  </EditableCodeBlockWrapper>
</Stack>`}
/>

> 

Reviewer: ```suggestion
## Composition notes

Ensure the surrounding page contains any information required to successfully use the editor. For example, include the expected language or input constraints.

Supporting content can be placed before or next to an Editable Code Block by using an introductory [Paragraph](/components/paragraph, [Callout](/components/callout), or [Card](/components/card).

```

## When to use Editable Code Block

<DoDont>
  <Do title="Do" body="Use Editable Code Block for a multi-line code input." />
  <Dont
    title="Don't"
    body="Don’t use Textarea for a multi-line code input. The monospace formatting and line numbers of Editable Code Block can help a user check that their code is entered correctly."
  />
</DoDont>

<DoDont>
  <Do
    title="Do"

>     body="Pair Editable Code Block with a visible label. This can be the Editable Code Block Header, a page Heading, or Label."

Reviewer 2: ```suggestion
    body="Pair Editable Code Block with a visible label. This can be the Editable Code Block Header, an associated page Heading, or Label."
```
